lib/route/cls/flower.c

```diff
@@ -0,0 +1,779 @@
+	        NLA_PUT_U8(msg, TCA_FLOWER_KEY_IP_TOS, f->cf_ip_dscp);
+
+	if (f->cf_mask & FLOWER_ATTR_IP_DSCP_MASK)
+	        NLA_PUT_U8(msg, TCA_FLOWER_KEY_IP_TOS_MASK, f->cf_ip_dscp_mask);
+
+        return 0;
+
+ nla_put_failure:
+	return -NLE_NOMEM;
+}
+
+static void flower_free_data(struct rtnl_tc *tc, void *data)
+{
+	struct rtnl_flower *f = data;
+
+	if (f->cf_act)
+		rtnl_act_put_all(&f->cf_act);
+
+	if (f->cf_police)
+		nl_data_free(f->cf_police);
+
+	if (f->cf_dst_mac)
+	        nl_data_free(f->cf_dst_mac);
+
+	if (f->cf_dst_mac_mask)
+	        nl_data_free(f->cf_dst_mac_mask);
+
+	if (f->cf_src_mac)
+	        nl_data_free(f->cf_src_mac);
+
+	if (f->cf_src_mac_mask)
+	        nl_data_free(f->cf_src_mac_mask);
+}
+
+static int flower_clone(void *_dst, void *_src)
+{
+        struct rtnl_flower *dst = _dst, *src = _src;
+
+	if (src->cf_dst_mac && !(dst->cf_dst_mac = nl_data_clone(src->cf_dst_mac)))
+	        return -NLE_NOMEM;
+
+	if (src->cf_dst_mac_mask &&
+	    !(dst->cf_dst_mac_mask = nl_data_clone(src->cf_dst_mac_mask)))
+	        return -NLE_NOMEM;
+
+	if (src->cf_src_mac && !(dst->cf_src_mac = nl_data_clone(src->cf_src_mac)))
+	        return -NLE_NOMEM;
+
+	if (src->cf_src_mac_mask &&
+	    !(dst->cf_src_mac_mask = nl_data_clone(src->cf_src_mac_mask)))
+	        return -NLE_NOMEM;
+
+	if (src->cf_act) {
+		if (!(dst->cf_act = rtnl_act_alloc()))
+			return -NLE_NOMEM;
+
+		memcpy(dst->cf_act, src->cf_act, sizeof(struct rtnl_act));
+
+		/* action nl list next and prev pointers must be updated */
+		nl_init_list_head(&dst->cf_act->ce_list);
+
+		if (src->cf_act->c_opts &&
+		    !(dst->cf_act->c_opts = nl_data_clone(src->cf_act->c_opts)))
+			return -NLE_NOMEM;
+
+		if (src->cf_act->c_xstats &&
+		    !(dst->cf_act->c_xstats = nl_data_clone(src->cf_act->c_xstats)))
+			return -NLE_NOMEM;
+
+		if (src->cf_act->c_subdata &&
+		    !(dst->cf_act->c_subdata = nl_data_clone(src->cf_act->c_subdata)))
+			return -NLE_NOMEM;
+
+		if (dst->cf_act->c_link) {
+			nl_object_get(OBJ_CAST(dst->cf_act->c_link));
+		}
+
+		dst->cf_act->a_next = NULL;   /* Only clone first in chain */
+	}
+
+	if (src->cf_police &&
+	    !(dst->cf_police = nl_data_clone(src->cf_police)))
+		        return -NLE_NOMEM;
+
+        return 0;
+}
+
+static void flower_dump_details(struct rtnl_tc *tc, void *data,
+				struct nl_dump_params *p)
+{
+	struct rtnl_flower *f = data;
+
+	if (!f)
+	        return;
+
+	if (f->cf_mask & FLOWER_ATTR_FLAGS)
+		nl_dump(p, " flags %u", f->cf_flags);
+
+	if (f->cf_mask & FLOWER_ATTR_PROTO)
+	        nl_dump(p, " protocol %u", f->cf_proto);
+
+	if (f->cf_mask & FLOWER_ATTR_VLAN_ID)
+	        nl_dump(p, " vlan_id %u", f->cf_vlan_id);
+
+	if (f->cf_mask & FLOWER_ATTR_VLAN_PRIO)
+	        nl_dump(p, " vlan_prio %u", f->cf_vlan_prio);
+
+	if (f->cf_mask & FLOWER_ATTR_VLAN_ETH_TYPE)
+	        nl_dump(p, " vlan_ethtype %u", f->cf_vlan_ethtype);
+
+	if (f->cf_mask & FLOWER_ATTR_DST_MAC)
+	        nl_dump(p, " dst_mac %s", f->cf_dst_mac);
```

Comment:
Here, these MAC addresses are `struct nl_data *` and cannot be printed this way. Above I suggested to make the fields arrays. Either way, the MAC addresses are binary and not NUL terminated. They cannot be printed this way. For example, link.c does:
```
nl_dump(p, "%s ", nl_addr2str(link->l_addr, buf, sizeof(buf)));
```
though, I don't think that the addresses should be `struct nl_addr` because their length is fixed. But something along the line.

Reply:
Oh, I missed that. Thx.